### ISSUE TITLE
still false positive for www.ad.nl

### ISSUE THREAD

**Issue by EPiC-APOC** [mmotti/pihole-regex#31]
It's mostly in reply of: https://github.com/mmotti/pihole-regex/commit/8eab5f81dd57bc5e9985ecc59f34111003312bf4

Exact match found in regex blacklist
   `^ad([sxv]?[0-9]*|system)[_.-]([^.[:space:]]+\.){1,}|^.+[_.-]ad([sxv]?[0-9]*|system)[_.-]`

I workaround this by whitelisting www.ad.nl, but I thought I'd mention it. Probably can't be done else I guess, so just keep or add the exception to the rule, right? 

With kind regards

**Comment by mmotti** [comment]
Good spot - I will look into this. I hadn't realised that www. was active
for ad.nl too

**Comment by mmotti** [comment]
OK so this is a bit of a tricky one.

It's a very unfortunate domain to have as a false positive and rather tricky one to work around without either making the regexp much less effective for the real "bad" domains or adding to the whitelist **www.ad.nl** to the whitelist.

Being that pi-hole implements POSIX ERE (or at least that was the case last time I looked), I'm also unable to add specific exceptions with things like negative lookaheads.

If I remove the regexp that allows 'ad' to be found anywhere after a period (not just at the start of the string), we'll miss domains such as:
```
zeus.ad.xiaomi.com
www.zeus.ad.xiaomi.com
```

If I combine the regexp to something like `^(.+[_.-])?ad([sxv]?[0-9]*|system)[_.-]([^.[:space:]]+\.){1,}`, it would permit **www.ad.nl** but steers away from the desired capture group and we would miss domains such as:

```
0.0.0.0 ad.a-ads.com
0.0.0.0 click.a-ads.com
0.0.0.0 static.a-ads.com
0.0.0.0 pokkt.fuse-ad.com
0.0.0.0 amazon-adsystem.com
0.0.0.0 ps-us.amazon-adsystem.com
0.0.0.0 s.adx.io
0.0.0.0 core.ad20.net
````

I could perhaps add a whitelist recommendation to the README.md and perhaps as a comment within the regex.list file itself. I'm not comfortable with programmatically forcing a whitelisted item upon everybody that runs the install script.

@DL6ER I'm sorry to bother you as I know you are very busy, but are you able to sanity check here and confirm that I'm not missing an obvious solution to this regexp wise other than providing a whitelist suggestion?

**Comment by badmojr** [comment]
Hey,
If you were to add a whitelist recommendation, do include **anti-ad.net**.
Thanks you for the regex list.
I use it in my blocklist project.

**Comment by mmotti** [comment]
Please see https://github.com/mmotti/pihole-regex/commit/1cd44d88be77a25fb1dc9410dbbe9a591cc7c5ad

It's not an ideal solution, but for now it's the best thing that I can think of without also enforcing a whitelist on people too.

In summary:

- Note in the README
- Note in the regex.list file itself
- Creation of whitelist.list (for manual reference)

**Comment by DL6ER** [comment]
> are you able to sanity check here and confirm that I'm not missing an obvious solution to this regexp wise other than providing a whitelist suggestion?

No, there cannot be such a solution with your current regex as the part

> `^.+[_.-]ad`

will just always match `]any characters].ad` which clearly matches the `www.ad.nl` by design. I'm also wondering why you are using `^.+` in the second path (precisely why you are enforcing at least one character in front of the `[_.-]`), but you will likely have your reasons for doing this and I'm not going to argue on this one.

> Being that pi-hole implements POSIX ERE (or at least that was the case last time I looked), I'm also unable to add specific exceptions with things like negative lookaheads.

That's correct. There are improvements on the way, however, negative lookaheads is not included as they are really complex to implement in conjunction with all the existing rules. POSIX doesn't define them. If you find an alternative implementation which we could use (both implementation- and license-wise), we could discuss this. Just mentioning: Things like embedding `perl` into FTL seems to be too much.

After all, I don't think negative lookahead should be used at all to embed indirect whitelisting through the regex filters themselves. Not only because this will quite significantly blow up the regex filters over time, but mainly because negative lookaheads are really complex regex-engine-wise. They have a quite dramatic impact on performance in my experience.

TL;DR You are doing everything right. Whitelisting an exception to an intentionally broad (and fast!) regex seems to be the best way.

**Comment by mmotti** [comment]
>I'm also wondering why you are using ^.+ in the second path (precisely why you are enforcing at least one character in front of the [_.-]), but you will likely have your reasons for doing this and I'm not going to argue on this one.

Honestly `.+` was purely out of habit on this occasion rather than for any specific reason, however I do believe the `^.+`  in its entirety is unnecessary in the case and more of an oversight on my part; I forgot to remove it whilst I was in the process of testing / splitting the original regexp apart to stop it from matching domains such as `ad.nl` and `ad.st`.

> Things like embedding perl into FTL seems to be too much.

I understand - I can appreciate that some regex implementations could also place unnecessary bulk to FTL and be more expensive with resources.

> After all, I don't think negative lookahead should be used at all to embed indirect whitelisting through the regex filters themselves.

I agree. The only times I could have really have seen it being helpful to me would to be able to specify to exclude a preceding **www.***, but other than that I wouldn't want to absolutely butcher a regexp lol.

--

In any case, I have removed the unnecessary requirements (`^.+`) and added several notes regarding potential false positives. I will see how many people have issues in future and if this false positives list grows considerably, I'll also consider adding an optional step during install to add these to the whitelist.

Thanks for taking the time reply - Appreciate the help and support as always.
